### PR TITLE
LRU cache for preventing duplicate writing instead of huge map.

### PR DESCRIPTION
**Pull request by KimKyungup**
## Proposed changes
This PR LRU cache for preventing duplicate writing.

This PR resolves the missing trie node during state migration.
This issue's root cause is skipping the trie node and its child when insertBlock added the trie node during migration.

This PR adds a cache for checking if the trie node is already written by only state migration, not insertBlock.
After this PR, trieSync check the cache instead of checking the target database.
So there is no issue to skip a trie.

We considered whole huge map. (https://github.com/klaytn/klaytn/pull/543) 
But it needs huge memory.
Therefore we applied LRU cache instead of the whold map.

This shows good result (lower memory, lower elapsed) comparing with whole map.
- whole map : 8h46m, 114GB
```
INFO[06/16,00:27:56 +09] [5] State migration is completed              committedCnt=363454052 elapsed=8h46m49.55906327s committed per second=11498.232
```

- LRU : 8h35m, 89.9GB
```
INFO[06/16,00:16:13 +09] [5] State migration is completed              committedCnt=363664086 elapsed=8h35m6.130689239s committed per second=11766.730
```

image

## Types of changes

Please put an x in the boxes related to your change.

- [ ] Bugfix
- [ ] New feature or enhancement
- [ ] Others

## Checklist

*Put an x in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.*

- [ ] I have read the [CONTRIBUTING GUIDELINES](https://github.com/klaytn/klaytn/blob/master/CONTRIBUTING.md) doc
- [ ] I have signed the [CLA](https://cla-assistant.io/klaytn/klaytn)
- [ ] Lint and unit tests pass locally with my changes (`$ make test`)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Related issues

- Please leave the issue numbers or links related to this PR here.

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: storage/statedb/sync.go [hunk 1]
```diff
@@ -117,9 +122,14 @@ func (s *TrieSync) AddSubTrie(root common.Hash, depth int, parent common.Hash, c
 		return
 	}
 	if s.bloom.Contains(root[:]) {
-		key := root.Bytes()
-		blob, _ := s.database.ReadStateTrieNode(key)
-		if local, err := decodeNode(key, blob); local != nil && err == nil {
+		// Bloom filter says this might be a duplicate, double check
+		if s.exist != nil {
+			if _, ok := s.exist.Get(root); ok {
```
Comment (ehnuje) [inline storage/statedb/sync.go:127]: Don't we need to check `s.database.HasStateTrieNode(root[:])` if `ok` is `false`? 
As `exist` is a cache, not a map.

Comment (KimKyungup) [reply storage/statedb/sync.go:127]: Right. The issue was the DB has a trie node which is not written by migration.
Therefore, this PR prevent to read DB. I will write more descriptions in this PR and explain the issue in offline.

Comment (ehnuje) [reply storage/statedb/sync.go:127]: Thanks for the explanation! How about leaving a comment that fast sync is the only case that checks the database?

Comment (KimKyungup) [reply storage/statedb/sync.go:127]: Thanks. I will add a comment. :)

**Comment by KimKyungup** [comment]
PTAL. :) @aidan-kwon @winnie-byun @ehnuje

**Comment by KimKyungup** [comment]
I will modify this code to use only lru cache, not bloom for migration.

**Comment by KimKyungup** [comment]
I reflected @aidan-kwon 's opinion that migration checks only lru cache without bloom.
But In fast sync, bloom filter will be used. So I leave the code.

File: storage/statedb/sync.go [hunk 2]
```diff
@@ -155,7 +163,13 @@ func (s *TrieSync) AddRawEntry(hash common.Hash, depth int, parent common.Hash)
 	if _, ok := s.membatch.batch[hash]; ok {
 		return
 	}
-	if s.bloom.Contains(hash[:]) {
+	if s.exist != nil {
+		if _, ok := s.exist.Get(hash); ok {
+			// already written in migration, skip the node
+			return
+		}
+	} else if s.bloom == nil || s.bloom.Contains(hash[:]) {
```
Comment (aidan-kwon) [inline storage/statedb/sync.go:171]:
```suggestion
	} else if s.bloom == nil && s.bloom.Contains(hash[:]) {
```

File: blockchain/state_migration.go [hunk 3]
```diff
@@ -99,14 +101,8 @@ func (bc *BlockChain) migrateState(rootHash common.Hash) error {
...
 
-	// TODO-Klaytn Change NewMemDB to real targetDB for restarting state migration
-	// Present bloom filter for migration.
-	// Since iterator doesn't support partitionedDB, we cannot use targetDB.
-	// If state migration is finished without restarting node, this fake empty DB is ok.
-	stateBloom := statedb.NewSyncBloom(uint64(512), database.NewMemDB())
-	defer stateBloom.Close()
-
-	trieSync := state.NewStateSync(rootHash, targetDB.DiskDB(), stateBloom)
+	lruCache, _ := lru.New(int(2 * units.Giga / common.HashLength)) // 2GB for 62,500,000 common.Hash key values
```
Comment (aidan-kwon) [inline blockchain/state_migration.go:104]:
```suggestion
	// NOTE: lruCache is medatory when state migration and block processing are executed simultaneously
	lruCache, _ := lru.New(int(2 * units.Giga / common.HashLength)) // 2GB for 62,500,000 common.Hash key values
```

File: blockchain/state_migration.go [hunk 4]
```diff
@@ -99,14 +101,9 @@ func (bc *BlockChain) migrateState(rootHash common.Hash) error {
...
 
-	// TODO-Klaytn Change NewMemDB to real targetDB for restarting state migration
-	// Present bloom filter for migration.
-	// Since iterator doesn't support partitionedDB, we cannot use targetDB.
-	// If state migration is finished without restarting node, this fake empty DB is ok.
-	stateBloom := statedb.NewSyncBloom(uint64(512), database.NewMemDB())
-	defer stateBloom.Close()
-
-	trieSync := state.NewStateSync(rootHash, targetDB.DiskDB(), stateBloom)
+	// NOTE: lruCache is medatory when state migration and block processing are executed simultaneously
```
Comment (winnie-byun) [inline blockchain/state_migration.go:104]:
```suggestion
	// NOTE: lruCache is mendatory when state migration and block processing are executed simultaneously
```

File: blockchain/state/sync.go [hunk 5]
```diff
@@ -22,14 +22,15 @@ package state
...
 import (
 	"bytes"
+	lru "github.com/hashicorp/golang-lru"
 	"github.com/klaytn/klaytn/blockchain/types/account"
 	"github.com/klaytn/klaytn/common"
 	"github.com/klaytn/klaytn/ser/rlp"
 	"github.com/klaytn/klaytn/storage/statedb"
 )
 
 // NewStateSync create a new state trie download scheduler.
```
Comment (winnie-byun) [inline blockchain/state/sync.go:32]:
```suggestion
// NewStateSync create a new state trie download scheduler.
// LRU cache is medatory when state syncing and block processing are executed simultaneously
```

File: storage/statedb/sync.go [hunk 6]
```diff
@@ -90,10 +91,11 @@ type TrieSync struct {
 	retrievedByDepth map[int]int              // Retrieved trie node number counted by depth
 	committedByDepth map[int]int              // Committed trie nodes number counted by depth
 	bloom            *SyncBloom               // Bloom filter for fast node existence checks
+	exist            *lru.Cache               // exist to check if the trie node is already written or not
 }
 
 // NewTrieSync creates a new trie data download scheduler.
```
Comment (winnie-byun) [inline storage/statedb/sync.go:97]:
```suggestion
// NewTrieSync creates a new trie data download scheduler.
// If both bloom and cache are set, only cache is used.
```

**Comment by KimKyungup** [comment]
PTAL. :)

**Comment by KimKyungup** [comment]
PTAL @ehnuje :)

**Pull request merged**